From 2bcff8a582216adbe2e32d9d1d5293558ec8ba04 Mon Sep 17 00:00:00 2001
From: Igor Grinberg <igrinberg@habana.ai>
Date: Sun, 15 Nov 2020 20:02:28 +0200
Subject: [PATCH] [SW-14676] fwupd: minimize wfc wait time for run

Rework the module_fwupd_sram_run() function to minimize the time spent
in wfc by moving all the work before wfc.
This is also in preparation for adding the security calls in this
routine.

Change-Id: Ia2fadb92a371c6d17d1b9cae197673cec735d8e3
Signed-off-by: Igor Grinberg <igrinberg@habana.ai>
---

diff --git a/preboot/module_fwupd.c b/preboot/module_fwupd.c
index 3d79b66..f25fee4 100644
--- a/preboot/module_fwupd.c
+++ b/preboot/module_fwupd.c
@@ -115,23 +115,31 @@
 
 static void module_fwupd_sram_run(void)
 {
+	int err;
+	enum hl_upd_sts sts;
+
 	module_fwupd_send_status(HL_UPD_STS_RUN_ACK, 0);
 
 	is_fit_checked = !fit_load_images();
-
-	if (module_fwupd_wfc(WFC_TIMEOUT_MS)) {
-		module_fwupd_send_status(HL_UPD_STS_TIMEOUT_ERR, 0);
-		preboot_puts("Timeout error\n");
-		return;
-	}
-
 	if (!is_fit_checked) {
-		module_fwupd_send_status(HL_UPD_STS_VALID_ERR, 0);
+		sts = HL_UPD_STS_VALID_ERR;
 		preboot_puts("Failed to load FIT image\n");
-		return;
+		goto fwupd_run_err;
 	}
+
+	err = module_fwupd_wfc(WFC_TIMEOUT_MS);
+	if (err)
+		goto fwupd_run_wfc_t;
 
 	module_fwupd_send_status(HL_UPD_STS_RUN_OK, 0);
+	return;
+
+fwupd_run_wfc_t:
+	sts = HL_UPD_STS_TIMEOUT_ERR;
+	preboot_puts("Timeout error\n");
+
+fwupd_run_err:
+	module_fwupd_send_status(sts, 0);
 }
 
 void module_fwupd_run(void)
